### PR TITLE
Ports VG's ComicIronic's awesome ventcrawling overhaul.

### PR DESCRIPTION
**Pull request by RemieRichards**
Ports https://github.com/d3athrow/vgstation13/pull/4004

@ComicIronic <3
- Ventcrawling now lets you walk through the pipe networks themselves, only popping out when you hit a vent.
- You can now vent crawl with Scrubbers.
- To balance the above, Scrubbers can now be welded, just like vents! (Shitty icon by myself!) Welded Scrubbers simply don't scrub, Like vents they may still receive signals, just not act upon them.

### CODE REVIEW

**Comment by palpatine213** [comment]
Can scrubbers be welded shut?

**Comment by RemieRichards** [comment]
No, but it really shouldn't be an issue given that vent crawling isn't a teleport anymore, AND it makes noise when somebody walks through a vent.

**Comment by Dorsisdwarf** [comment]
But this means we need a way to open fire on the piping system, breaching the pipe but shooting anything in it.

"THEY'RE IN THE VENTS! BLAST THE WALLS!"

**Comment by palpatine213** [comment]
Umm, this means that Xenos can Go ANYWHERE via ventcrawling. the method of securing an area via welding no longer works.

**Comment by RemieRichards** [comment]
@palpatine213 I really think it's a non issue, both the queen and sentinels are already an all access card.

**Comment by GunHog** [comment]
RR, this *IS* a major issue. We *need* to be able to secure these areas. It takes forever for Xeno acid to burn though things, and it is very obvious, which gives everyone plenty of time to evacuate or prepare. I do not think it a good idea to leave scubbers unweldable in this case. Although the easier option would be to only allow crawling through vents, which avoids this problem altogether.

**Comment by RemieRichards** [comment]
Welp, Time to enjoy 1/2 of the station's pipes being worthless to this feature.
If you want to secure the area what's stopping you unwrenching the pipes?

**Comment by Dorsisdwarf** [comment]
Just let people fucking weld scrubbers

**Comment by RemieRichards** [comment]
Oi, dont get pissy at me, I'm already working on it.

**Comment by Alek2ander** [comment]
It should hide the surface when you're in the vents.

**Comment by RemieRichards** [comment]
Weldable Scrubbers coming up, Shitty icon by myself included.

**Comment by as334** [comment]
Does it use the air from the pipes as the mobs air for breathing/fire and shit? because i really want to light distro loop on fire in order to stop an alien invasion

**Comment by RemieRichards** [comment]
Well, you're IN the pipe, so your atmos is fetched from the pipe.
Depends how pipe code handles it's gases "inside" pipes.

**Comment by optimumtact** [comment]
holy fucking shit you people aliens are not meant to be balanced at all.

**Comment by RemieRichards** [comment]
It's fine, Welded scrubbers are pretty cool anyways, now atmos techs can weld those too to further their floods.

**Comment by RemieRichards** [comment]
The latest commit marks the loss of a fallen comrade.
Goodbye "You weld the vent shit." we will miss you.

**Comment by Anonus** [comment]
Technically a massive nerf to ventcrawling, as the pipe system is barely logical at all. But it's actually pretty fun, up until someone fills the vents with superheated N2O.

**Comment by ComicIronic** [comment]
That depends on your perspective. This lets you see where you're going before you pop out, which is a far cry from "Oh there's a singulo here, who know".

**Comment by RemieRichards** [comment]
plus you can pray to the admins to let you be a pipe hobo.

**Comment by xxalpha** [comment]
what happens if the pipes are filled with hot burning gas? what happens if the pipes are super pressurized? what happens if the pipes have plasma in them and the crawler can't breathe plasma?

**Comment by RemieRichards** [comment]
as I said before, Depends how pipes deal with return_air() and other such "help me I'm in an object and need to breathe" procs.

**Comment by xxalpha** [comment]
so you could fuck over all aliens by pumping lava hot gas into the pipes?

**Comment by RemieRichards** [comment]
yet again the answer is predicated on how pipes work, I'm tired, I'm not looking at pipeatmos code.

**Comment by WJohn** [comment]
:+1: 
Though personally I wish it was all moved to the disposals system. It'd greatly reduce the amount of places they can go to and make a lot more sense in terms of what they can fit inside of.

**Comment by xxalpha** [comment]
and what's the movement speed while vent crawling? It should probably be at least slightly faster.

and this sound should occasionally play when something is vent crawling
http://hydra-media.cursecdn.com/dota2.gamepedia.com/a/a9/Weav_move_08.mp3

**Comment by Profakos** [comment]
.On Sayustation we have a similar thing. The person was inside the pipe, and indeed, all the gases affected it, meaning that monkeys all suffocated from low pressure, while xenos were living fine. We solved it by putting a check for being in pipes for handle_environment, which isn't a really neat solution at all.

**Comment by Incoming5643** [comment]
It'd be nice if mobs autocrawled forward until they reached a juncture where, if the player didn't manually select a choice quickly, a random selection moved them forward. This could both make the pipes more navigable and stymie camping in the pipenet.

**Comment by Stickymayhem** [comment]
Does this mean alt clicking a vent instantly lets you climb in because that would be great.

**Comment by Fox-McCloud** [comment]
:-1: 

Don't think this is good for balance.

It also encourages flooding distribution and scrubbers, both, with extremely high pressure CO2 to burn+incap changelings+xenos both "just in case".

**Comment by PKPenguin321** [comment]
>Does this mean alt clicking a vent instantly lets you climb in because that would be great.

this this this this
i've died as the first alien larva and a changeling monkey before because of the fucking popup menu
i'm 800% down for this if it makes you instantly jump into the vent

**Comment by WJohn** [comment]
@Fox-McCloud how is that a bad thing? You'd also be flooding everything else and get banned for it unless everyone HAPPENS to have welded everything shut, and in that case you'd also have already prevented xenos from using the vents or scrubbers at all. Besides, there's nothing preventing anyone from doing this already by just burning the shit out of everything as a traitor and coincidentally putting a stop to the xenos.

**Comment by Fox-McCloud** [comment]
provided no depressurization occurs in the station, the vents will never flood into the main station, as pressure in the vents is held back by an external pressure check.

**Comment by KorPhaeron** [comment]
I'm worried this is just going to make ventcrawl even more obnoxious. It's already basically a teleport with no cooldown, now you can camp inside the pipes as well

**Comment by Anonus** [comment]
Yeah but this means it's not an instant teleport, and camping in the pipes means you can't attack anybody until you pop out.

**Comment by Tyrranatar** [comment]
I think it should be made so that if you're in a pipe you can't see the mobs outside of it. It would allow some neat ambushes to be setup, be just a little more realistic, and make it harder to camp in them.

**Comment by WJohn** [comment]
This actually has some rather unique tactical advantages now. If you have the aliens cornered in a location with no alternative places to vent crawl, you could in theory cut the pipes leading up to them to sever them from the station remotely, rather than doing it at the source. :+1: 

Now if only pipes didn't prevent you from cutting them because of stupid pressure bullshit.

**Comment by KorPhaeron** [comment]
Tearing up pipes to fight xenomorphs sounds cool, but I'm more concerned that completely destroying the atmos system will now be a requirement to ever kill a changeling

**Comment by Iamgoofball** [comment]
Shitty idea:
what if T-Ray Scanners when active showed people in pipes along with the rest of the shit it normally shows

**Comment by xxalpha** [comment]
Yeah, T-Ray scanners detecting crawlers would be a neat buff to them.

**Comment by RemieRichards** [comment]
@Stickymayhem @PKPenguin321 it takes 25 ticks, or 2.5 seconds, to enter the vent which is fair I believe.
It's still infinitely faster than going from a vent at the top of the list to one at the bottom with that damn popup.

**Comment by pudl** [comment]
i hate it

**Comment by RemieRichards** [comment]
You hate everything.

**Comment by pudl** [comment]
i don't hate you :heart:

**Comment by Miauw62** [comment]
a lot of atmos techs will pump a LOT of pressure into the distro loop, so competent atmos techs could basically kill monkey ventcrawling.

**Comment by Anonus** [comment]
Even if you're putting basically all of the air tank into distro the highest it'll get is around 300-400 kPa.
Which isn't super lethal if you don't sit in it for too long.

**Comment by Miauw62** [comment]
i've gotten to about 4000 kpa before. (admitted, i was flooding plasma at this point of time, but i had already pumped a lot of custom airmix into distro, and once i started pumping plasma i disabled N2)

**Comment by Anonus** [comment]
For the most part, anything that will make the pipes lethal to be in will also make the station lethal to be in. It's also worth noting that even areas with no breaches get some air from the vents.

**Comment by GunHog** [comment]
Vents only activate by default if air pressure goes below 101.something kPa, or their settings are manually changed. Flooding the atmos pipes with normal air is common for some Atmos techs, as it means areas can be refilled in a decent amount of time. Due to this, the station can remain perfectly safe while the pipe's air would crush anything inside.

**Comment by Anonus** [comment]
Yeah, but even if there aren't any breaches, areas will sometimes get air from the vents; a good way to test this is to fill distro with plasma but change no other settings, eventually, even if there are no breaches or atmos changes, the plasma will leak out.

**Comment by Miauw62** [comment]
they dont dump out all of their contents though, the case im talking about
is literally just high pressure distro, which is common and useful.

**Comment by xxalpha** [comment]
Yeah monkey ventcrawling would be a thing of the past with this.

**Comment by Alek2ander** [comment]
Underfloor crawling?

**Comment by RemieRichards** [comment]
At worst it'll SHORTEN your vent escape journeys, for example on Box, Security --> Engineering would be certain death, but Perma Brig --> Outside Security, you'll likely make it (Given they're not actively trying to flush you out with non-standard pressures (flooding))

Granted the main users of this (Xenos and Drones) are immune to pressure damage, it's probably fine.

**Comment by Dorsisdwarf** [comment]
The perma brig has a seperate atmos network, you can only get to the courtoom.

**Comment by RemieRichards** [comment]
IT IS A THEORETICAL, LEAVE ME ALONE.

**Comment by astralenigma** [comment]
And from the courtroom you can get to the outside.

**Comment by AnturK** [comment]
:+1: 

Also the vent clog event should now pop all vent crawling critters out or do something silly to them

**Comment by Tokiko1** [comment]
What if you are a changeling and turn back human while inside a pipe?

**Comment by WJohn** [comment]
Or for that matter, a xeno spitting someone out while in the pipe, or a human bursting while in one, or a xeno getting reverse chest-bursted.

**Comment by RemieRichards** [comment]
Depends on those two scenarios, and if an unintended effect comes from either (People being ending up outside of the pipes, etc.)

Their loc will either be set to the turf of where the first mob (Monkeyling/Xeno) OR the exact loc of the first mob, So you'll either be in the pipe or out of it.

**Comment by RemieRichards** [comment]
@tgstation/commit-access paid my gold subscription, merge pls.

**Comment by GunHog** [comment]
Cheri told me you had Platinum!

**Comment by RemieRichards** [comment]
Must have been a loyalty reward!

File: code/ATMOSPHERICS/atmospherics.dm [hunk 1]
```diff
@@ -21,9 +21,19 @@ Pipelines + Other Objects -> Pipe network
...
+	var/welded = 0 //Used on pumps and scrubbers
+
 	var/global/list/iconsetids = list()
 	var/global/list/pipeimages = list()
 
+
+/obj/machinery/atmospherics/Destroy()
+	for(var/mob/living/L in src)
+		L.remove_ventcrawl()
+		L.loc = get_turf(src)
```
Comment (ComicIronic) [inline code/ATMOSPHERICS/atmospherics.dm]: One thing I just realised is that this doesn't called the Enter procs.
You might want to use forceMove ( or whatever the TG equivalent is )

Comment (RemieRichards) [inline code/ATMOSPHERICS/atmospherics.dm]: Ah yes, true.
Also yes, we have foreceMove()

Comment (RemieRichards) [inline code/ATMOSPHERICS/atmospherics.dm]: Also, thanks.

**Comment by RemieRichards** [comment]
Conflict fix + Changelog (@Jordie0608)

@tgstation/commit-access

**Comment by GunHog** [comment]
Poor RR. Even with your status, you still have the debuff that makes all your stuff take longer to get merged!

File: code/modules/mob/living/ventcrawling.dm [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+	for(var/atom/A in totalMembers)
+		var/image/new_image = image(A, A.loc, dir = A.dir)
+		pipes_shown += new_image
+		if(client)
+			client.images += new_image
+
+
+/mob/living/proc/remove_ventcrawl()
+	for(var/image/current_image in pipes_shown)
+		client.images -= current_image
```
Comment (phil235) [inline code/modules/mob/living/ventcrawling.dm]: needs an if(client) check above to avoid runtime

**Comment by RemieRichards** [comment]
@phil235 remove_ventcrawl() should be safer now.

**Comment by RemieRichards** [comment]
@phil235 <3

**Comment by norill** [comment]
you should be able to see all pipes (distro, waste, mixing, ect), not just the loop you entered, bc loops are connected

**Comment by RemieRichards** [comment]
@norill you specifically only see the pipes connected to the vent/scrubber you entered.
Meaning if you can see that pipe, you can move to it.
if you can't see a pipe, you can't move to it from that pipe network.

**Comment by norill** [comment]
but you can crawl through pumps meaning you can go from distro to waste via pump in atmos. or basically everywhere

**Comment by astralenigma** [comment]
You know that the only part where the scrubs and vents connect is atmos meaning that to get to a scrubber in the same room through a vent on that room you would have to go to atmosia first and then come back.
Honestly due to the pairings is preferable to just use one of them.

**Comment by RemieRichards** [comment]
Yeah I've noticed the actual issue now, You can use atmos to go from vents to scrubbers, without being able to see the scrubbers, I'll think of a fix/talk with Comic.

**Pull request merged**